### PR TITLE
Resolve Mailer TODOs & warnings, fixing port data in README 

### PR DESCRIPTION
**Pull request by alfabravoteam**
On Mailer class:
+ Solved a TODO left in a previous commit on the SSL connection setup to SMTP server. So, `socketFactory` property was replaced by `mail.smtp.ssl.enable`. Additionally, added another property, `mail.smtp.ssl.checkserveridentity` in the SSL set-up, to enforce server identity check to avoid MITM attacks.
+ Also, addressed older TODOs about `Jenkins.getInstance` using the method recommended by Jenkins [docs](https://javadoc.jenkins-ci.org/jenkins/model/Jenkins.html#getInstanceOrNull--) (`getInstanceOrNull`).
+ A couple of `@deprecated` annotations were on top of javaDoc blocks, hence they were being ignored. Moved them on top of the declaration and checked that javaDoc worked as expected.
+ Addressed a TODO about a try-with-resources for a BulkChange instance that was not properly closed in case of an exception.

Changed text in the README.adoc about the TLS default port (was wrong) and the SMTP port value depending on the configuration to make it clearer.

Changed MailerTest to assertNotNull instead of assertTrue on a not null validation.

Moved a repeated string in MimeMessageBuilder into a final const to use it were required

### CODE REVIEW

File: src/test/java/jenkins/plugins/mailer/MailerJCasCCompatibilityTest.java [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
 import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
 
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+
+import hudson.tasks.Mailer;
+import io.jenkins.plugins.casc.misc.RoundTripAbstractTest;
+
```
Comment (alecharp) [inline src/test/java/jenkins/plugins/mailer/MailerJCasCCompatibilityTest.java:11]: Import reorder could be mis-interpreted. Not a blocker here, as it's a test class.

File: src/main/java/hudson/tasks/Mailer.java [hunk 2]
```diff
@@ -411,7 +408,8 @@ private static Session createSession(String smtpHost, String smtpPort, boolean u
         }
 
         private static Authenticator getAuthenticator(final String smtpAuthUserName, final String smtpAuthPassword) {
-            if(smtpAuthUserName==null)    return null;
+            if(smtpAuthUserName==null)
+            	return null;
```
Comment (alecharp) [inline src/main/java/hudson/tasks/Mailer.java:412]:
```suggestion
            if(smtpAuthUserName == null) {
            	return null;
            }
```

File: src/main/java/hudson/tasks/Mailer.java [hunk 3]
```diff
@@ -371,16 +370,14 @@ private static Session createSession(String smtpHost, String smtpPort, boolean u
             	 * and thats done in mail sender, and it would be a bit of a hack to get it all to
             	 * coordinate, and we can make it work through setting mail.smtp properties.
             	 */
-            	if (props.getProperty(SMTP_SOCKETFACTORY_PORT_PROPERTY) == null) {
```
[the remark this answers is not in the document]
Comment (alfabravoteam) [reply src/main/java/hudson/tasks/Mailer.java:374]: Understood! :)

File: src/test/java/jenkins/plugins/mailer/MailerJCasCCompatibilityTest.java (hunk 1 above)
Comment (alfabravoteam) [reply src/test/java/jenkins/plugins/mailer/MailerJCasCCompatibilityTest.java:11]: I moved from my work's laptop with IntelliJ to my personal laptop with Eclipse and I forgot not to mess with things unnecessarily. Sorry.

**Comment by alfabravoteam** [comment]
> 
> 
> Could you please review the suggestions I made? Thanks.

Thanks for double checking things. All suggestions made sense as it seemed more of me overusing IDE features and forgetting not to move things if they're working.

File: src/main/java/hudson/tasks/Mailer.java [hunk 4]
```diff
@@ -371,16 +370,14 @@ private static Session createSession(String smtpHost, String smtpPort, boolean u
...
+            	if (props.getProperty(SMTP_SOCKETFACTORY_PORT_PROPERTY)==null) {
```
Comment (alecharp) [inline src/main/java/hudson/tasks/Mailer.java:373]: Still, you removed the spaces around the `==`

```suggestion
            	if (props.getProperty(SMTP_SOCKETFACTORY_PORT_PROPERTY) == null) {
```

File: src/main/java/hudson/tasks/Mailer.java [hunk 5]
```diff
@@ -371,16 +370,14 @@ private static Session createSession(String smtpHost, String smtpPort, boolean u
...
                     String port = smtpPort==null?"465":smtpPort;
                     props.put(SMTP_PORT_PROPERTY, port);
                     props.put(SMTP_SOCKETFACTORY_PORT_PROPERTY, port);
             	}
-            	if (props.getProperty("mail.smtp.socketFactory.class") == null) {
-                    // TODO SocketFactory properties are now discouraged. Using (`mail.smtp.ssl.enable`, "true") should suffice
-                    // https://javaee.github.io/javamail/FAQ#commonmistakes
-            		// Example also found at https://javaee.github.io/javamail/FAQ#smtpssl
-                    props.put("mail.smtp.socketFactory.class","javax.net.ssl.SSLSocketFactory");
+            	if (props.getProperty(SMTP_SSL_ENABLE_PROPERTY)==null) {
```
Comment (alecharp) [inline src/main/java/hudson/tasks/Mailer.java:378]:
```suggestion
            	if (props.getProperty(SMTP_SSL_ENABLE_PROPERTY) == null) {
```

**Review by alecharp — CHANGES_REQUESTED** [review]
please see my suggestions

File: src/main/java/hudson/tasks/Mailer.java [hunk 6]
```diff
@@ -237,7 +237,11 @@ public static InternetAddress StringToAddress(String strAddress, String charset)
     public static DescriptorImpl DESCRIPTOR;
 
     public static DescriptorImpl descriptor() {
-        return Jenkins.get().getDescriptorByType(Mailer.DescriptorImpl.class);
+        final Jenkins jenkins = Jenkins.getInstanceOrNull();
+        if (jenkins == null) {
+            throw new IllegalStateException("Jenkins instance is not ready");
+        }
+        return jenkins.getDescriptorByType(Mailer.DescriptorImpl.class);
```
Comment (oleg-nenashev) [inline src/main/java/hudson/tasks/Mailer.java:244]:
```suggestion
        return Jenkins.get().getDescriptorByType(Mailer.DescriptorImpl.class);
```

Messed up the merge conflict resolution

**Review by alecharp — APPROVED** [review]
This seems to be good now. Thank you @alfabravoteam

**Review by oleg-nenashev — APPROVED** [review]
All changes look good to me

**Pull request merged**